compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala

```diff
@@ -0,0 +1,186 @@
+import NullOpsDecorator._
+
+/** This module defines methods to interpret types of Java symbols, which are implicitly nullable in Java,
+ *  as Scala types, which are explicitly nullable.
+ *
+ *  The transformation is (conceptually) a function `n` that adheres to the following rules:
+ *    (1) n(T)              = T|JavaNull              if T is a reference type
+ *    (2) n(T)              = T                       if T is a value type
+ *    (3) n(C[T])           = C[T]|JavaNull           if C is Java-defined
+ *    (4) n(C[T])           = C[n(T)]|JavaNull        if C is Scala-defined
+ *    (5) n(A|B)            = n(A)|n(B)|JavaNull
+ *    (6) n(A&B)            = n(A) & n(B)
+ *    (7) n((A1, ..., Am)R) = (n(A1), ..., n(Am))n(R) for a method with arguments (A1, ..., Am) and return type R
+ *    (8) n(T)              = T                       otherwise
+ *
+ *   Treatment of generics (rules 3 and 4):
+ *     - if `C` is Java-defined, then `n(C[T]) = C[T]|JavaNull`. That is, we don't recurse
+ *       on the type argument, and only add JavaNull on the outside. This is because
+ *       `C` itself will be nullified, and in particular so will be usages of `C`'s type argument within C's body.
+ *       e.g. calling `get` on a `java.util.List[String]` already returns `String|Null` and not `String`, so
+ *       we don't need to write `java.util.List[String|Null]`.
+ *     - if `C` is Scala-defined, however, then we want `n(C[T]) = C[n(T)]|JavaNull`. This is because
+ *       `C` won't be nullified, so we need to indicate that its type argument is nullable.
+ *
+ *   Notice that since the transformation is only applied to types attached to Java symbols, it doesn't need
+ *   to handle the full spectrum of Scala types. Additionally, some kinds of symbols like constructors and
+ *   enum instances get special treatment.
+ */
+object JavaNullInterop {
+
+  /** Transforms the type `tp` of Java member `sym` to be explicitly nullable.
+   *  `tp` is needed because the type inside `sym` might not be set when this method is called.
+   *
+   *  e.g. given a Java method
+   *  String foo(String arg) { return arg; }
+   *
+   *  After calling `nullifyMember`, Scala will see the method as
+   *
+   *  def foo(arg: String|JavaNull): String|JavaNull
+   *
+   *  This nullability function uses `JavaNull` instead of vanilla `Null`, for usability.
+   *  This means that we can select on the return of `foo`:
+   *
+   *  val len = foo("hello").length
+   *
+   *  But the selection can throw an NPE if the returned value is `null`.
+   */
+  def nullifyMember(sym: Symbol, tp: Type)(implicit ctx: Context): Type = {
+    assert(ctx.explicitNulls)
+    assert(sym.is(JavaDefined), "can only nullify java-defined members")
+
+    // A list of "policies" that special-case certain members.
+    // The policies should be disjoint: we use the first one that is applicable.
+    val whitelist: Seq[NullifyPolicy] = Seq(
```

Comment:
We don't really have the luxury to write code like this in the compiler. Creating a Seq only to immediately take it apart is too wasteful, unless the Seq can be created at an outer level. Even then it's better to use  a concrete type (I would recommend `List` since the sequence is short). 

Look around how highly optimized ClassfileParser is. You are calling this code from the middle of it!

Reply:
Whitelist is removed

Reply:
@noti0na1 refactored JavaNullInterop so that it's a lot more efficient now. All the data objects are now gone, and are replaced by hardcoded if checks. Done.